From 10b57c0f65031b8b9c84d5061a8804d1fed1bfad Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Fri, 20 Oct 2017 20:00:26 -0400
Subject: [PATCH] Optimize ItemStack.isEmpty()

Remove hashMap lookup every check, simplify code to remove ternary

diff --git a/src/main/java/net/minecraft/item/ItemStack.java b/src/main/java/net/minecraft/item/ItemStack.java
index 33438ef..d75f8da 100644
--- a/src/main/java/net/minecraft/item/ItemStack.java
+++ b/src/main/java/net/minecraft/item/ItemStack.java
@@ -182,13 +182,15 @@ public final class ItemStack {
         this.updateEmptyState();
     }
 
+    // Paper start - optimize isEmpty
+    private static Item airItem;
     public boolean isEmpty() {
-        return this == EMPTY
-                ? true
-                : (this.item != null && this.item != Item.getItemFromBlock(Blocks.AIR)
-                        ? (this.stackSize <= 0 ? true : this.itemDamage < -32768 || this.itemDamage > 65535)
-                        : true);
+        if (airItem == null) {
+            airItem = Item.REGISTRY.getObject(new ResourceLocation("air"));
+        }
+        return this == EMPTY || (this.item == null || this.item == airItem || (this.stackSize <= 0 || (this.itemDamage < -32768 || this.itemDamage > 65535)));
     }
+    // Paper end
 
     public static void registerFixes(DataFixer fixer) {
         fixer.registerWalker(FixTypes.ITEM_INSTANCE, new BlockEntityTag());
-- 
2.18.0

